core/services/blockheaderfeeder/block_header_feeder.go

```diff
@@ -0,0 +1,247 @@
+	if minBlockNumber == nil {
+		lggr.Debug("no blocks to store")
+		return nil
+	}
+
+	earliestStoredBlockNumber, err := f.findEarliestBlockNumberWithBlockhash(ctx, lggr, minBlockNumber.Uint64()+1, uint64(toBlock))
+	if err != nil {
+		return errors.Wrap(err, "finding earliest blocknumber with blockhash")
+	}
+
+	if earliestStoredBlockNumber == nil {
+		// store earliest blockhash and return
+		// on next iteration, earliestStoredBlockNumber will be found and
+		// will make progress in storing blockhashes using blockheader
+		f.bhs.StoreEarliest(ctx)
+		lggr.Info("Stored earliest block number")
+		return nil
+	}
+
+	blocks, err := blockhashstore.DecreasingBlockRange(earliestStoredBlockNumber.Sub(earliestStoredBlockNumber, big.NewInt(1)), minBlockNumber)
+	if err != nil {
+		return err
+	}
+
+	for i := 0; i < len(blocks); i += int(f.storeBlockhashesBatchSize) {
+		j := i + int(f.storeBlockhashesBatchSize)
+		if j > len(blocks) {
+			j = len(blocks)
+		}
+		blockRange := blocks[i:j]
+		blockHeaders, err := f.getRlpHeadersBatch(ctx, blockRange)
+		if err != nil {
+			return errors.Wrap(err, "fetching block headers")
+		}
+
+		lggr.Debugw("storing block headers", "blockRange", blockRange)
+		err = f.batchBHS.StoreVerifyHeader(ctx, blockRange, blockHeaders)
+		if err != nil {
+			return errors.Wrap(err, "storing block headers")
+		}
+	}
+
+	if f.lastRunBlock != 0 {
+		// Prune stored, anything older than fromBlock can be discarded
+		for block := f.lastRunBlock - uint64(f.lookbackBlocks); block < uint64(fromBlock); block++ {
+			if _, ok := f.stored[block]; ok {
+				delete(f.stored, block)
+				f.lggr.Debugw("Pruned block from stored cache",
+					"block", block, "latestBlock", latestBlock)
+			}
+		}
+	}
+	f.lastRunBlock = latestBlock
+	return nil
+}
+
+func (f *BlockHeaderFeeder) findLowestBlockNumberWithoutBlockhash(ctx context.Context, lggr logger.Logger, blockToRequests map[uint64]map[string]struct{}) *big.Int {
+	var min *big.Int
+	for block, unfulfilledReqs := range blockToRequests {
+		if len(unfulfilledReqs) == 0 {
+			continue
+		}
+		if _, ok := f.stored[block]; ok {
+			// Already stored
+			continue
+		}
+		stored, err := f.bhs.IsStored(ctx, block)
+		if err != nil {
+			lggr.Warnw("Failed to check if block is already stored",
+				"error", err,
+				"block", block)
+		} else if stored {
+			lggr.Infow("Blockhash already stored",
+				"block", block, "unfulfilledReqIDs", blockhashstore.LimitReqIDs(unfulfilledReqs))
+			f.stored[block] = struct{}{}
+			continue
+		}
+		blockNumber := big.NewInt(0).SetUint64(block)
+		if min == nil || min.Cmp(blockNumber) >= 0 {
+			min = blockNumber
+		}
+	}
+	return min
+}
+
+func (f *BlockHeaderFeeder) findEarliestBlockNumberWithBlockhash(ctx context.Context, lggr logger.Logger, startBlock, toBlock uint64) (*big.Int, error) {
+	from := startBlock
+	for i := startBlock; i < toBlock; i += uint64(f.getBlockhashesBatchSize) {
```

Comment:
So this loop range is `[startBlock, toBlock)` (i.e `toBlock` is not included)? I guess that's fine, since `toBlock` will be in the next range that is scanned by the feeder?

Reply:
yeah next scan should catch the `toBlock` (assuming that `latestHead` advances)